---
title: "Sweetener community analysis"
author: "Vinita Periwal"
date: "`r Sys.Date()`"
output: 
  
  html_document:
    code_folding: hide
    fig_caption: true
    toc: true
    toc_float:
          collapsed: false
          smooth_scroll: false
  editor_options:
    chunk_output_type: console
---

<h4><i>Sweeteners 16S community sequencing</i></h4>
<h4><i></i></h4>

<style>
p.caption {
  font-size: 1.2em;
}

.table caption {
    font-size: 1.2em;
}
</style>


##loading packages

loading the required packages.

```{r echo=TRUE, setup, include=FALSE}

knitr::opts_chunk$set(echo = TRUE,
                      warning = FALSE,
                      message = FALSE,
                      cache = FALSE,
                      tidy = FALSE,
                      size = "small",
                      fig.width = 10,
                      fig.height = 8
)

library(microbiome)
library(phyloseq)
library(DESeq2)
library(vegan)
library(data.table)
library(dplyr)
library(tibble)
library(reshape2)
library(janitor) # removes empty rows and columns
library(tidyr)
library(stringr)
library(RColorBrewer)
library(ggplot2)
library(lattice)
library(ggsci) #fancy color palettes
library(Cairo) #high resolution images
library(ggforce) #facet wrapping
library(ggpubr) #reg equation
library(car) #for Boxplots with labels
library(gtools) #mixedsort
library(data.table)
library(readxl)
library(lme4)
library(jtools) #for p-values and table representation of lmer objects
library(broom.mixed) #for tidying lmer objects
library(ggeffects)
library(emmeans)
library(knitr) #kable for tidy output in RMD used in for loop
#library(kableExtra) #scrolling kable tables
library(upstartr)
library(hrbrthemes)
library(ape)
library(microbiomeutilities)
library(eulerr)
library(microbial)
library(metacoder)
library(EnhancedVolcano)
library(dada2)
library(pheatmap)
library(lme4)
library(SummarizedExperiment)
library(lefser)
library(microViz)
library(ggbreak)

#global theme
th<-theme(plot.title = element_text(size = 12, face = "bold"),axis.title=element_text(size=12,color = "black"),
          axis.text.x = element_text(size=12, color = "black"),axis.text.y = element_text(size=12, color = "black"))

```
# Objective
In the primary screen the combination of duloxetine and isosteviol showed a synergistic effect therefore in the current experiment the impact of the two compounds is tested on community level.

The community was assembled using the 25 bugs used in primary screening assay and labelled as:

1. InoMix - inoculation mix (starter mix) - passage 0
2. Mix1 - community treated with the test compounds from passage 1 - 5
3. Mix2 - community treated with test compounds after 3 passages and continued to 8 passages.

We explore changes in microbial diversity in presence of duloxetine, isosteviol and duloxetine-isosteviol passaged over time. Initial analysis to derive the abundances and taxa tables were done by Dada2 processing pipeline (Dada2_processing.Rmd and details in methods of manuscript). The statistical analysis of the ASVs obtained is performed here.

### data

Create phyloseq object

```{r, echo=TRUE}
setwd("~/ShikiFactory/WP3/SW_SF100/Community/sweet_community/")
#reading data
#metadata
MetaDF<-data.frame(read.table(file = 'metadata_sampbiol.tsv', header = TRUE, sep = '\t'))
rownames(MetaDF)<-MetaDF$sample.id
MetaDF<-MetaDF[,2:10]

#OTU table
otumat<-data.frame(read.table(file = "ASV_25bugs.tsv", sep = '\t', header =  TRUE))
rownames(otumat)<-otumat$ASV
otumat<-data.matrix(otumat[,2:177])
#head(otumat)

#Taxa table
taxamat<-data.frame(read.table(file = "TAXA_25bugs.tsv", sep = '\t', header =  TRUE))
rownames(taxamat)<-taxamat$ASV
taxamat<-as.matrix(taxamat[,2:8])
#head(taxamat)

#convert all data to phyloseq objects
OTU<-otu_table(otumat, taxa_are_rows = TRUE)
TAX<-tax_table(taxamat)
Samp<-sample_data(MetaDF)
#sample_names(Samp)

#create phyloseq object
physeq<-phyloseq(OTU,TAX,Samp)
physeq

# we will merge technical replicates now
merge_physeq<-merge_samples(physeq, "samp_biolrep")
sample_names(merge_physeq)

# once merged, we need to recreate a phyloseq object with new sample names
#read merged metadata 
MetaDF1<-data.frame(read.table(file = 'metadata.tsv', header = TRUE, sep = '\t'))
#head(MetaDF1)
rownames(MetaDF1)<-MetaDF1$sample.id
MetaDF1<-MetaDF1[,2:8]
#head(MetaDF1)

#convert all data to phyloseq objects
OTU1<-otu_table(merge_physeq, taxa_are_rows = TRUE)
TAX1<-tax_table(taxamat)
Samp1<-sample_data(MetaDF1)
Samp1

final_physeq<-phyloseq(OTU1, TAX1, Samp1)
sample_names(final_physeq)

#convert to df
physeq_df<-psmelt2(final_physeq)
colnames(physeq_df)[3]<-'Abundance'
colnames(physeq_df)[2]<-'SampleID'
View(physeq_df)

##Check key data variables
#number of samples
nsamples(final_physeq) #total 88

#number of taxa
ntaxa(final_physeq) #998

#total number of reads in the dataset
reads_per_OTU <- taxa_sums(final_physeq)
print(sum(reads_per_OTU)) #7736099

##checking sequencing depth: plot
sdt = data.table(as(sample_data(final_physeq), "data.frame"),
                 TotalReads = sample_sums(final_physeq), keep.rownames = TRUE)
setnames(sdt, "rn", "SampleID")
#head(sdt)
ggplot(sdt, aes(TotalReads, fill=compound)) + geom_histogram(color='black',linewidth=0.1) + ggtitle("Sequencing Depth") + th + theme_bw() + geom_vline(xintercept=30000, color= "red", linetype='dashed') + scale_fill_nejm() 

#subset samples
ino<-subset_samples(final_physeq, type == 'Ino-Mix')
ino

mix1<-subset_samples(final_physeq, type == 'Mix1')
mix1

mix2<-subset_samples(final_physeq, type == 'Mix2')
mix2

#plot of community composition at inoculation
#Inoculation Mix
pseq_ino <- ino %>%
  aggregate_taxa(level = "Phylum") %>%  
  microbiome::transform(transform = "compositional")
pseq_ino

CairoSVG(file = "fig2a", width = 7, height = 4, bg = "white")
plot_composition(pseq_ino,
                      taxonomic.level = "Phylum",
                      sample.sort = "biol_rep",
                      x.label = "biol_rep",
                      otu.sort = "abundance") +
  scale_fill_cosmic() +
  labs(x = "Biological replicates", y = "Relative abundance",
       title = "InoMix: Relative abundance data") + theme_classic() + 
  theme(legend.text = element_text(size=16), legend.title = element_text(size=18)) +
  guides(fill=guide_legend(title="Phylum")) + th
dev.off()

#inoMix
pseq_ino <- ino %>%
  aggregate_taxa(level = "Species") %>%  
  microbiome::transform(transform = "compositional")
pseq_ino_df<-psmelt(pseq_ino)
class(pseq_ino_df$passage)<-'character'
pseq_ino_df<-pseq_ino_df[,c(4,3,6,8,9,12:17)]
View(pseq_ino_df)

CairoSVG(file = "figIno", width = 20, height = 20, bg = "white")
pseq_ino_df %>%
  ggplot(aes(x=biol_rep, y=Abundance)) + geom_boxplot(lwd=0.3) + geom_jitter(aes(color=biol_rep)) + facet_wrap(~Species) + theme_bw() +
  scale_color_cosmic()
dev.off()


#Mix1
pseq_mix1 <- mix1 %>%
  aggregate_taxa(level = "Species") %>%  
  microbiome::transform(transform = "compositional")
pseq_mix1_df<-psmelt(pseq_mix1)
class(pseq_mix1_df$passage)<-'character'
pseq_mix1_df<-pseq_mix1_df[,c(4,3,6,8,9,12:17)]
View(pseq_mix1_df)


compare_means(Abundance ~ compound,  data = pseq_mix1_df)
my_comparisons <- list( c("iso-steviol-duloxetine", "duloxetine"), c("iso-steviol-duloxetine", "iso-steviol"), c("iso-steviol-duloxetine", "DMSO"), c("iso-steviol","duloxetine"), c("duloxetine", "DMSO"), c("iso-steviol", "DMSO") )

CairoSVG(file = "figMix1", width = 20, height = 17, bg = "white")
pseq_mix1_df %>%
  ggplot(aes(x=compound, y=Abundance)) + geom_boxplot(lwd=0.3) + geom_jitter(aes(color=passage)) + facet_wrap(~Species, scales = "free_y") + theme_bw() + stat_compare_means(comparisons = my_comparisons) +
  scale_color_cosmic()
dev.off()


# DMSO vs iso-steviol-duloxetine
s<-pseq_mix1_df %>% filter(compound %in% c("DMSO","iso-steviol-duloxetine") & biol_rep == "Schedule_1" & passage == "5")
s<-s[,c(11,3,2)]
dmso<-s %>% filter(compound == 'DMSO')
iso<- s %>% filter(compound != 'DMSO')
new_s<-data.frame(merge(dmso,iso, by="Species"))
View(new_s)


new_s %>% ggplot(aes(x=Abundance.x, y=Abundance.y)) + geom_point(aes(color=Species),size=3, alpha=0.8) + geom_abline() +
  scale_color_igv() + theme_minimal() + th + scale_y_break(breaks = c(0.2,0.4), scales = 0.5) + theme(axis.text.x = element_text(size=18, color = "black"),axis.text.y = element_text(size=18, color = "black"), legend.text = element_text(size = 14))



#Schedule 1
CairoSVG(file = "fig2D", width = 15, height = 8, bg = "white")
pseq_mix1_df %>% filter(biol_rep == "Schedule_1") %>%
  ggplot(aes(x=passage, y=Abundance)) + geom_bar(color = "black", lwd=0.2, stat = "identity", position = "fill", aes(fill=Species),) + theme_bw() + scale_fill_igv() + facet_grid(~compound) + theme(axis.text.x = element_text(size=14, color = "black"),axis.text.y = element_text(size=14, color = "black"), legend.text = element_text(size = 14), legend.position = "right", legend.justification = "center", strip.text.x = element_text(size = 14)) + th
dev.off()

#Schedule 2
pseq_mix1_df %>% filter(biol_rep == "Schedule_2") %>%
  ggplot(aes(x=passage, y=Abundance)) + geom_bar(color = "black", lwd=0.2, stat = "identity", position = "fill", aes(fill=Species),) + theme_bw() + scale_fill_igv() + facet_grid(~compound) + theme() + th



# #schedule 1 and passage 5
# subset_samples(pseq_mix1, passage == "5" & biol_rep == "Schedule_1") %>%
#   comp_barplot("Species", n_taxa = 25, merge_other = FALSE, label = NULL,
#                bar_outline_colour = "black",
#                order_with_all_taxa = TRUE) +
#   facet_grid(c("compound","biol_rep"), scales = "free") + # scales = "free" is IMPORTANT!
#   coord_flip() +
#   theme(axis.ticks.y = element_blank(), strip.text = element_text(face = "bold"))
# 
# #schedule 2 and passage 5
# subset_samples(pseq_mix1, passage == "5") %>%
#   comp_barplot("Species", n_taxa = 25, merge_other = FALSE, label = NULL,
#                bar_outline_colour = "black",
#                order_with_all_taxa = TRUE, facet_by = "biol_rep") +
#   facet_grid(c("compound","biol_rep"), scales = "free") + # scales = "free" is IMPORTANT!
#   coord_flip() +
#   theme(axis.ticks.y = element_blank(), strip.text = element_text(face = "bold"))


# set.seed(50)
# subset_samples(pseq_mix1, passage == "5" & biol_rep == "Schedule_1") %>%
#   tax_transform("compositional", rank = "Species") %>%
#   comp_heatmap(tax_anno = taxAnnotation(
#     Prev. = anno_tax_prev(bar_width = 0.3)), 
#     sample_anno = sampleAnnotation(
#       CMP = anno_sample("sample"),
#       border = FALSE),
#       colors = heat_palette(palette = "PurpOr", rev = TRUE),
#     grid_col = "black"
#     )
# 
# 
# subset_samples(pseq_mix1, passage == "5" & biol_rep == "Schedule_2") %>%
#   tax_transform("compositional", rank = "Species") %>%
#   comp_heatmap(tax_anno = taxAnnotation(
#     Prev. = anno_tax_prev(bar_width = 0.3)), 
#     sample_anno = sampleAnnotation(
#       CMP = anno_sample("sample"),
#       border = FALSE),
#       colors = heat_palette(palette = "PurpOr", rev = TRUE),
#     grid_col = "black"
#     )


#Mix2
pseq_mix2 <- mix2 %>%
  aggregate_taxa(level = "Species") %>%  
  microbiome::transform(transform = "compositional")
pseq_mix2_df<-psmelt(pseq_mix2)
pseq_mix2_df<-pseq_mix2_df[,c(4,3,6,8,9,12:17)]
pseq_mix2_df<-pseq_mix2_df %>% filter(compound != 'mGAM only')
View(pseq_mix2_df)

compare_means(Abundance ~ compound,  data = pseq_mix2_df)
my_comparisons <- list( c("iso-steviol-duloxetine", "duloxetine"), c("iso-steviol-duloxetine", "iso-steviol"), c("iso-steviol-duloxetine", "DMSO"), c("iso-steviol","duloxetine"), c("duloxetine", "DMSO"), c("iso-steviol", "DMSO") )

CairoSVG(file = "figcompMix2", width = 20, height = 20, bg = "white")
pseq_mix2_df %>%
  ggplot(aes(x=compound, y=Abundance)) + geom_boxplot(lwd=0.3) + geom_jitter(aes(color=as.character(passage))) + facet_wrap(~Species, scales = "free_y") + theme_bw() + stat_compare_means(comparisons = my_comparisons) +
  scale_color_cosmic()
dev.off()

plot_composition(pseq_mix2,
                      taxonomic.level = "Species",
                      sample.sort = "biol_rep",
                      x.label = "sample") +
  scale_fill_d3(palette = "category20b") +
  guides(fill = guide_legend(ncol = 1)) +
  labs(x = "Samples", y = "Relative abundance",
       title = "Mix2: Relative abundance data") + 
  theme_ipsum(grid="Y") +
  theme(legend.text = element_text(face = "italic"), axis.text.x = element_text(angle = 90,hjust = 1)) + 
  guides(fill=guide_legend(title="Species")) + th

set.seed(123)
#Relative abundance taxa of mix1 at genus level
pseq_mix1 <- mix1 %>%
  aggregate_taxa(level = "Species") %>%
  microbiome::transform(transform = "compositional")

#Core taxa shared 
# simple way to count number of samples in each group
table(meta(pseq_mix1)$compound, useNA = "always")
treatments <- unique(as.character(meta(pseq_mix1)$compound))
print(treatments)
runs <- unique(as.character(meta(pseq_mix1)$passage))
print(runs)

pseq_mix1_sub<-subset_samples(pseq_mix1, biol_rep == "Schedule_1")
otu_table(pseq_mix1_sub)

list_core <- data.frame(stringsAsFactors = FALSE) # an empty object to store information

for (n in treatments){ # for each variable n in compounds
    #print(paste0("Identifying Core Taxa for ", n))
    
    ps.sub <- subset_samples(pseq_mix1_sub, treatments == n)
    
    for(i in 1:length(runs)){
    
    samp<-subset_samples(ps.sub, passage == runs[i])
    samp
    
    core_m <- core_members(samp, detection = 0.000001, prevalence = 1, include.lowest = TRUE) 
    print(paste0("No. of core taxa in Mix-1 ", n, "passage", i," : ",  length(core_m))) 
    core<-cbind("Compound" = n, "Passage" = i, "core_num" = length(core_m), "Core_sp" = core_m, "abund" = value)
    list_core<-rbind(list_core, core)
    
    }
    
}

View(list_core)
str(list_core)
list_core$comp_psg<-paste0(list_core$Compound,"_",list_core$Passage)

pal<-paletteer_d("colorBlindness::SteppedSequential5Steps")
CairoSVG(file = "fig2E", width = 10, height = 6, bg = "white")
ggplot(list_core, aes(comp_psg,Core_sp)) + geom_tile(aes(fill = comp_psg), color="white") + 
   scale_fill_discrete(type=pal) + theme_classic() + theme(axis.text.x = element_blank(), axis.text.y = element_text(size=14, color = "black",face = "italic"), legend.text = element_text(size = 14))
dev.off()

list_core1 <- c() # an empty object to store information

for (n in treatments){ # for each variable n in compounds
    #print(paste0("Identifying Core Taxa for ", n))
    
    ps.sub <- subset_samples(pseq_mix2, treatments == n) # Choose sample from compound by n
    core_m <- core_members(ps.sub, detection = 0.0001, include.lowest = TRUE) # ps.sub is phyloseq selected with only samples from g 

    print(paste0("No. of core taxa in Mix-2 ", n," : ", length(core_m))) # print core taxa identified in each treatment.
    list_core1[[n]] <- core_m # add to a list core taxa for each group.
    #print(list_core)
    
}

print(list_core1)
plot(venn(list_core1))

list_core2 <- c() # an empty object to store information

core_m <- core_members(pseq_ino, detection = 0.0001, include.lowest = TRUE)

print(paste0("No. of core taxa in InoMix "," : ", length(core_m))) # print core taxa
core_m # core taxa for each group.


```

### Alpha Diversity
Alpha diversity measures the diversity within a single sample and is generally based on the number and relative abundance of taxa at some rank (e.g. species or OTUs).

Compute and compare all common alpha diversity metrics: "Observed", "Chao1", "Shannon", "Simpson", "InvSimpson", "Fisher"

Significance testing: performed using two-way repeated measures ANOVA and if significant (p<0.05) perform posthoc Tukey-test for multiple pairwise comparisons

```{r, echo=TRUE}
#alpha diversity
alpha_div <-data.frame(microbiome::alpha(final_physeq, index = "all"))
#head(alpha_div)
alpha_div<-tibble::rownames_to_column(alpha_div, "SampleID")
nrow(alpha_div)
p.shannon <- boxplot_alpha(final_physeq, 
                           index = "shannon",
                           x_var = "type")

p.shannon + theme_bw() + scale_fill_nejm() + facet_wrap(~biol_rep) + th +
  labs(x="Samples", y="Shannon diversity") + ggtitle("Alpha Diversity")


#alpha diversity over time for exposure and established

#merge data frames
df<-data.frame(merge(alpha_div, sdt, by = 'SampleID'))
#head(df)
nrow(df)

#Mix1
CairoSVG(file = "fig2b", width = 6, height = 3, bg = "white")
df %>% filter(type=='Mix1') %>% ggplot(aes(x=passage,y=diversity_shannon)) + geom_point(aes(color=compound)) + geom_smooth(aes(fill=compound,color=compound), alpha=0.05) +
  scale_color_cosmic() + theme_bw() + scale_fill_cosmic() + theme(legend.text = element_text(size=14), legend.title = element_text(size=16)) +
labs(x="Passage", y="Shannon diversity") + th + ggtitle("Alpha diversity over passages: Mix1")
dev.off()

#Mix2
CairoSVG(file = "fig2c", width = 6, height = 3, bg = "white")
df %>% filter(type=='Mix2' & compound != 'mGAM only') %>% ggplot(aes(x=passage,y=diversity_shannon)) + geom_point(aes(color=compound)) + geom_smooth(aes(fill=compound,color=compound), alpha=0.05) + scale_color_cosmic() + scale_fill_cosmic() + theme_bw() + theme(legend.text = element_text(size=14), legend.title = element_text(size=16)) +
labs(x="Passage", y="Shannon diversity") + th + ggtitle("Alpha diversity over passages: Mix2")
dev.off()

#plot all diversity measures (Mix1)
plot_richness(mix1, x='compound', nrow = 2, measures = c("Observed", "Chao1", "Shannon", "Simpson", "InvSimpson", "Fisher")) + geom_boxplot(aes(fill=compound), alpha=0.5) + 
scale_fill_npg() + theme_bw() + th + theme(axis.text.x = element_blank(),axis.ticks.x = element_blank()) + ggtitle("Diversity measures: Mix1")

#plot all diversity measures (Mix2)
plot_richness(mix2, x='compound', nrow = 2, measures = c("Observed", "Chao1", "Shannon", "Simpson", "InvSimpson", "Fisher")) + geom_boxplot(aes(fill=compound), alpha=0.5) + 
scale_fill_npg() + theme_bw() + th + theme(axis.text.x = element_blank(),axis.ticks.x = element_blank()) + ggtitle("Diversity measures: Mix2")

#Mix 1
df1<-df %>% filter(type == 'Mix1')

###analysis of variance (Mix1)
class(df1$passage)<-'character'
m <- aov(diversity_shannon ~ compound + passage, data=df1)
summary(m)

#estimated marginal means (least-squares means) Within group differences compared in every phase
em<-emmeans(m, specs = pairwise~compound)
em

eff_size(em, sigma = sigma(m), edf = 10)

#head(data.frame(em$contrasts %>% summary(infer = TRUE)))
plot(em, comparisons = TRUE)

#pairwise p-value plot (uses 'Tukey' adjusted p-value)
CairoSVG(file = 'Fig2bii', width = 6, height = 3, bg = 'white')
pwpp(em$emmeans) + scale_color_cosmic() + theme_minimal() + ggtitle("Mix1: pairwise comparison p-value plot")
dev.off()


#Mix 2
df2<-df %>% filter(type == 'Mix2' & treatment == 'exposure')

###analysis of variance (Mix2)
class(df2$passage)<-'character'
m <- aov(diversity_shannon ~ compound + passage, data=df2)
summary(m)

#estimated marginal means (least-squares means) Within group differences compared in every phase
em<-emmeans(m, specs = pairwise~compound)
em

eff_size(em, sigma = sigma(m), edf = 10)

#head(data.frame(em$contrasts %>% summary(infer = TRUE)))
plot(em, comparisons = TRUE)

#pairwise p-value plot (uses 'Tukey' adjusted p-value)
CairoSVG(file = 'Fig2cii', width = 6, height = 3, bg = 'white')
pwpp(em$emmeans) + scale_color_cosmic() + theme_minimal() + ggtitle("Mix2: pairwise comparison p-value plot")
dev.off()

```

### Beta Diversity
Beta diversity is a way to quantify the difference between two communities. The typical way beta diversity is plotted is using ordination. Ordination is a way to display “high dimensional” data in a viable number of dimensions (2 to 3). Our data is “high dimensional” because we have many samples with many species and species can be considered a “dimension”. Since this is mock community where species are known to us thus we don't compute indexes with presence or absence of data.

Indexes used with count data: bray curtis, weighted unifrac)

1. bray curtis: the sum of lesser counts for species present in both communities divided by the sum of all counts in both communities.
2. Weighted Unifrac: The fraction of the phylogenetic tree branch lengths shared by the two communities, weighted by the counts of organisms, so more abundant organisms have a greater influence.

PS: rarefaction is not used because it throws away 90% of data

Results: Compounds have a significant effect on the community. But is it really possible or is it a chance event. Checked beta dispersity, the dispersity is not significant means adonis results are good.

```{r, echo=TRUE}

#beta diversity
#weighted unifrac distance comparison
k<-rowSums(otu_table(pseq_mix1)) > 0 #remove empty zero rows
bt<-otu_table(pseq_mix1)[k,]
beta_dist<-vegdist(t(bt), method = 'bray', binary = TRUE, na.rm = TRUE)
#beta_dist
mds <- metaMDS(beta_dist)
mds$points
mds_data <- as.data.frame(mds$points)
#head(mds_data)
mds_data$SampleID <- rownames(mds_data)
mds_data<-data.frame(mds_data, row.names = NULL)

#head(df1)
mds_data<-merge(mds_data, df1)
#class(mds_data$passage)<-"character"
#head(mds_data)
ggplot(mds_data, aes(x = MDS1, y = MDS2, color = as.factor(passage))) +
  geom_point() + scale_color_jco() + theme_bw() + stat_ellipse(aes(group = passage), linetype = 2) + ggtitle("Bray-curtis dissimilarity") + th

#ordination plot
# Ordinate the data
set.seed(4235421)
ord <- ordinate(pseq_mix1, "PCoA", "bray")

plot_ordination(pseq_mix1, ord, color = "compound", shape = 'biol_rep') +
  geom_point(size = 5) + theme_bw() + scale_color_npg() + 
  ggtitle("Bray-curtis relative abundance") + th +
  stat_ellipse(aes(group = compound), linetype = 2)

#PERMANOVA for community-level multivariate comparisons
pseq.rel <- microbiome::transform(pseq_mix1, "compositional")
otu <- abundances(pseq.rel)
meta <- meta(pseq.rel)

permanova <- adonis(t(otu) ~ compound,
               data = meta, permutations=99, method = "bray")
# P-value
print(as.data.frame(permanova$aov.tab)["compound", "Pr(>F)"])


```

### Differential abundance analysis - DESeq2

Pairwise comparison to highlight significant taxa abundance between treated (iso-steviol-duloxetine) and control group (DMSO) in Mix1. Log2FC > 1.5 and a wald-test p-value < 0.05 is considered significant. The difference of abundant taxa is plotted as genus.

```{r, echo=TRUE}
sample_data(mix1)$compound<-relevel(as.factor(sample_data(mix1)$compound), "DMSO")
sample_data(mix1)$compound<-as.factor(sample_data(mix1)$compound)

ds2 <- phyloseq_to_deseq2(mix1, ~ compound)
sample_data(mix1)
# Run DESeq2 analysis (all taxa at once!)
dds <- DESeq(ds2)
dds

# Investigate results
res1 <- results(dds, contrast = c('compound','iso-steviol-duloxetine','DMSO'))
#res1
deseq.results <- as.data.frame(res1)
deseq.results
df <- deseq.results
df$taxon <- rownames(df)
df <- df %>% arrange(log2FoldChange, padj)
df

df$id<-rownames(df)
taxo<-data.frame(tax_table(mix1))
taxo$id<-rownames(taxo)
#head(taxo)

taxnames_deseq.results<-merge(df, taxo, by='id')
#head(taxnames_deseq.results)


EnhancedVolcano(taxnames_deseq.results,
                lab = taxnames_deseq.results$Species,
                x = 'log2FoldChange',
                y = 'padj',
                pCutoff = 0.05,
                FCcutoff = 1,
                legendPosition = 'right',
                legendLabels=c('not sig.',bquote(log[2]~'fold change'),'p-value',
                               bquote('p-value &' ~log[2]~'fold change')),
                labSize = 3.5,
                pointSize = 2,
                col=c('#737373', '#02818a', '#fec44f', '#67001f'),
                colAlpha = 0.7,
                xlim = c(-35,35),
                legendLabSize = 12,
                legendIconSize = 3.0,
                axisLabSize = 12,
                title = "Treatment: Iso-steviol-duloxetine, Control: DMSO",
                subtitle = "",
                caption = ""
                )


```
